File: 11403_경로찾기/path.js
/*

2018 Jan 29, made by Moon

*/
var fs = require('fs');
var input = fs.readFileSync('./txt.txt').toString().split('\n');
var N = parseInt(input[0]);
input.shift();
var ix, ixLen, jx, jxLen, pathInfo = [], rowData;
for(ix = 0, ixLen = input.length; ix < ixLen; ix++){
    pathInfo[ix] = [];
    rowData = input[ix].split(' ');
    for(jx = 0, jxLen = rowData.length; jx < jxLen; jx++){
        pathInfo[ix].push(parseInt(rowData[jx]));
    }
}

var graph = [];
var path = [];
var visited = [];

function initGraph(){
    var ix, ixLen, jx, jxLen;
    for(ix = 0, ixLen = pathInfo.length; ix < ixLen; ix++){
        graph[ix] = pathInfo[ix].slice();
    }

    for(ix = 0, ixLen = N; ix < ixLen; ix++){
        path[ix] = [];
        for(jx = 0, jxLen = N; jx < jxLen; jx++){
            path[ix][jx] = 0;
        }
    }

    initVisited();
}

function dfs(curNode, head, isStart) {
    var ix, ixLen;

    if(!isStart) {
        visited[curNode] = true;
        path[head][curNode] = 1;
    }

    for (ix = 0, ixLen = graph[curNode].length; ix < ixLen; ix++) {
        if(graph[curNode][ix] !== 0 && !visited[ix]){
            dfs(ix, head);
        }
    }
}

function initVisited() {
    var ix, ixLen;
    for (ix = 0, ixLen = N; ix < ixLen; ix++) {
        visited[ix] = false;
    }
}

function dfsAll() {
    var i, ixLen, head;
    for (i = 0, ixLen = graph.length; i < ixLen; i++) {
        head = i;
        dfs(i, head, true);
        initVisited();
    }
}

function printPath(){
    var ix, ixLen, result;
    for (ix = 0, ixLen = path.length; ix < ixLen; ix++) {
        result = path[ix].join(' ')
        console.log("%s", result);
    }
}

function main(){
    initGraph();
    dfsAll();
    printPath();
}

main();

File: 11403_경로찾기/path.cpp
/*

2018 Feb 02, made by Moon

*/
#include <iostream>

using namespace std;

int N;
int graph[101][101];
int result[101][101];
int visited[101];

void dfs(int head, int root) {
	visited[root]++;
	for (int i = 1; i <= N; i++) {
		if (!visited[i] && graph[root][i]) {
			result[head][i] = 1;
			dfs(head, i);
		}
	}
}

void dfsAll() {
	for (int i = 1; i <= N; i++) {
		visited[i] = -1;
		dfs(i, i);
        for(int j=1; j<=N; j++){
            visited[j] = 0;
        }
	}
}

int main() {
	cin >> N;

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= N; j++) {
			cin >> graph[i][j];
		}
	}


	dfsAll();
	//출력
	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= N; j++) {
			cout << result[i][j] << " ";
		}
		cout << endl;
	}

	return 0;
}
